src/plugins/intel_gna/src/gna_infer_request.cpp

```diff
@@ -0,0 +1,174 @@
+// Copyright (C) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "gna_infer_request.hpp"
+
+#include "gna_plugin.hpp"
+
+namespace GNAPluginNS {
+
+GNAInferRequest::GNAInferRequest(const std::shared_ptr<GNAPlugin>& plg,
+                                 const std::vector<std::shared_ptr<const ov::Node>>& inputs,
+                                 const std::vector<std::shared_ptr<const ov::Node>>& outputs)
+    : InferenceEngine::IInferRequestInternal(inputs, outputs),
+      plg(plg) {
+    CreateInferRequest();
+}
+
+GNAInferRequest::GNAInferRequest(const std::shared_ptr<GNAPlugin>& plg,
+                                 InferenceEngine::InputsDataMap networkInputs,
+                                 InferenceEngine::OutputsDataMap networkOutputs)
+    : InferenceEngine::IInferRequestInternal(networkInputs, networkOutputs),
+      plg(plg) {
+    CreateInferRequest();
+}
+
+void GNAInferRequest::InferImpl() {
+    // execute input pre-processing.
+    execDataPreprocessing(_inputs);
+    // result returned from sync infer wait method
+
+    bool result = false;
+    auto infer_call = [&]() {
+        result = plg->Infer(_inputs, _outputs);
+    };
+
+    CallCleanupAndRethrowOnException(std::move(infer_call));
+
+    // if result is false we are dealing with QoS feature and set kRequestIndexInvalid
+    // if result is ok we set kRequestIndexCompleted to not execute request if it is not
+    // in the queue.
+    auto result_request_index = result ? kRequestIndexCompleted : kRequestIndexInvalid;
+    SetRequestIndex(result_request_index);
```

Comment:
Maybe it is better to put it inside `infer_call`? Because in case of exception you set request id

Reply:
I am not sure if get your comment properly, because I think in case of exception request id will be set to invalid and exception will be rethrown.
Due the fact infer_call is executed in context of 
```
void GNAInferRequest::CallCleanupAndRethrowOnException(std::function<void()>&& function_to_invoke) {
    try {
        function_to_invoke();
    } catch (...) {
        // need to preserve invalid state here to avoid next Wait() from clearing it
        // and next rethrow issue.
        SetRequestIndex(kRequestIndexInvalid);
        throw;
    }
}
```

Reply:
> I am not sure if get your comment properly, because I think in case of exception request id will be set to invalid and exception will be rethrown. Due the fact `infer_call` is executed in context of
> 
> ```
> void GNAInferRequest::CallCleanupAndRethrowOnException(std::function<void()>&& function_to_invoke) {
>     try {
>         function_to_invoke();
>     } catch (...) {
>         // need to preserve invalid state here to avoid next Wait() from clearing it
>         // and next rethrow issue.
>         SetRequestIndex(kRequestIndexInvalid);
>         throw;
>     }
> }
> ```

Yes, but there is no sense to keep it outside `infer_call`.

Reply:
You comment is right, but on the other hand I wanted to make this lambdas as simple as possible and consistent across infer/start_async. Intention was to do there only calls to `plg`.

Reply:
> You comment is right, but on the other hand I wanted to make this lambdas as simple as possible and consistent across infer/start_async. Intention was to do there only calls to `plg`.

for `queue_call` it is not true, look unaligned